### PR TITLE
skia: image: optimization: faster drawing of fully opaque images

### PR DESCRIPTION
**Pull request by flsobral**
## Description:
Optimizes drawing fully opaque images by copying the image pixels directly to the canvas.
Images that are not fully opaque or that require any transformation must still use drawBitmapRect.

Makes drawing JPEG and opaque PNGs over 10x faster when running on CPU. 🚀 

### Related Issue:
#63 

## Motivation and Context:
Poor performance using PathAnimation on iMX6ULL

## Benefited Devices:
Any system with graphical stack based on Skia, which right now are:
 - Linux (X86_64/ARM);
 - Android;
 - iOS;
 - MacOSX

## How Has This Been Tested?
Tests performed using MyTapp sample prototype and TotalCross Showcase

## Tested Devices:
 - iMX6ULL
 - MacOSX

### CODE REVIEW

File: TotalCrossVM/src/nm/ui/android/skia.cpp [hunk 1]
```diff
@@ -222,7 +222,11 @@ int skia_makeBitmap(int32 id, void *data, int32 w, int32 h) {
 
     if (id < 0) { // must create a new bitmap
         SkBitmap bitmap;
-        bitmap.installPixels(SkImageInfo::Make(w, h, kN32_SkColorType, kUnpremul_SkAlphaType), (void*)converted, sizeof(Pixel) * w, releaseProc, nullptr);
+		bitmap.installPixels(SkImageInfo::Make(w, h, kN32_SkColorType, kUnpremul_SkAlphaType),
```
Comment (acmlira) [inline TotalCrossVM/src/nm/ui/android/skia.cpp:225]: Unrelated changes.

**Comment by Farshid-Monhaseri** [comment]
> In which cases the non-native code will ever run?

For now, in no case, the  input format is always ARGB, and it is not supported by Skia, and we need to convert it to a supported one like BGRA. It was done by manual shifting and I used compiler byte swap to do this operation efficiently. But in my point of view it is only a temporary workaround, and we should change the drawing strategy to avoid such this extra operations and overheads.

File: TotalCrossVM/src/nm/ui/android/skia.cpp (hunk 1 above)
Comment (flsobral) [reply TotalCrossVM/src/nm/ui/android/skia.cpp:286]: Those are actually related! The releaseProc was changed to be able to also handle freeing dstPixels, which is allocated as a void* instead of an integer type.

File: TotalCrossVM/src/nm/ui/android/skia.cpp [hunk 2]
```diff
@@ -251,6 +252,20 @@ int skia_makeBitmap(int32 id, void *data, int32 w, int32 h) {
         SkBitmap bitmap;
 		bitmap.installPixels(SkImageInfo::Make(w, h, kN32_SkColorType, kUnpremul_SkAlphaType),
 							 (void*)converted, sizeof(Pixel) * w, releaseProc, nullptr);
+
+#if USE_COLORTYPE_CONVERSION
+        void* dstPixels = operator new(w * h * canvas->imageInfo().bytesPerPixel());
```
[the remark this answers is not in the document]
Comment (flsobral) [reply TotalCrossVM/src/nm/ui/android/skia.cpp]: The first conversion swaps ARGB to BGRA I believe, Skia doesn't support ARGB directly.
The second allocation does not necessarily have the same size of the first one, because it uses the system's pixel format, it will usually be smaller, but it may be bigger? Not sure.

This second array is used by bitmap.readPixel, which performs the conversion from BGRA to the system's pixel format (RGB888, RGB565 or whatever).
Then we install this second array into the original bitmap, along with all the updated info. The install calls the releaseProc, which frees the first array - I double checked that, we're not leaking memory here. ;)

So for now, we do need both conversions, but later I'm thinking about changing the Image object to use BGRA instead of ARGB, which would allow us to drop the first conversion.

File: TotalCrossVM/src/nm/ui/android/skia.cpp [hunk 3]
```diff
@@ -234,7 +235,7 @@ int32 skia_stringWidth(const void *text, int32 charCount, int32 typefaceIndex, i
 
 static void releaseProc(void* addr, void* ) {
     // printf("releaseProc called\n");
-    delete[] (uint32_t*) addr;
+    operator delete (addr);
```
[the remark this answers is not in the document]
Comment (flsobral) [reply TotalCrossVM/src/nm/ui/android/skia.cpp:250]: Will be dealt later by issue #289

**Review by teras — APPROVED** [review]
